### PR TITLE
catch parse error from corrupt config

### LINKED ISSUE #45316: Flutter Unhandled exception:  FormatException: Unterminated string (at line 2, character 24)   "name": "PullToDismis
## Steps to Reproduce

I installed flutter on my Mac mini running Catalina all the commands yield the same result

## Logs

Unhandled exception:
FormatException: Unterminated string (at line 2, character 24)
  "name": "PullToDismis
                       ^

#0      _ChunkedJsonParser.fail (dart:convert-patch/convert_patch.dart:1394:5)
#1      _ChunkedJsonParser.close (dart:convert-patch/convert_patch.dart:505:9)
#2      _parseJson (dart:convert-patch/convert_patch.dart:32:10)
#3      JsonDecoder.convert (dart:convert/json.dart:495:36)
#4      JsonCodec.decode (dart:convert/json.dart:153:41)
#5      new Config (package:flutter_tools/src/base/config.dart:14:41)
#6      runInContext.<anonymous closure> (package:flutter_tools/src/context_runner.dart:84:21)
#7      AppContext._generateIfNecessary.<anonymous closure> (package:flutter_tools/src/base/context.dart:100:41)
#8      _LinkedHashMapMixin.putIfAbsent (dart:collection-patch/compact_hash.dart:293:23)
#9      AppContext._generateIfNecessary (package:flutter_tools/src/base/context.dart:88:20)
#10     AppContext.get (package:flutter_tools/src/base/context.dart:117:32)
#11     Config.instance (package:flutter_tools/src/base/config.dart:18:41)
#12     new _DefaultUsage.<anonymous closure> (package:flutter_tools/src/reporting/usage.dart:212:25)
#13     WhereIterator.moveNext (dart:_internal/iterable.dart:442:11)
#14     MappedIterator.moveNext (dart:_internal/iterable.dart:394:19)
#15     Iterable.join (dart:core/iterable.dart:364:19)
#16     new _DefaultUsage (package:flutter_tools/src/reporting/usage.dart:215:10)
#17     new Usage (package:flutter_tools/src/reporting/usage.dart:79:9)
#18     runInContext.<anonymous closure> (package:flutter_tools/src/context_runner.dart:119:20)
#19     AppContext._generateIfNecessary.<anonymous closure> (package:flutter_tools/src/base/context.dart:100:41)
#20     _LinkedHashMapMixin.putIfAbsent (dart:collection-patch/compact_hash.dart:293:23)
#21     AppContext._generateIfNecessary (package:flutter_tools/src/base/context.dart:88:20)
#22     AppContext.get (package:flutter_tools/src/base/context.dart:117:32)
#23     Usage.instance (package:flutter_tools/src/reporting/usage.dart:85:40)
#24     flutterUsage (package:flutter_tools/src/reporting/usage.dart:69:33)
#25     _handleToolError (package:flutter_tools/runner.dart:123:5)
#26     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:77:13)
#27     _rootRunBinary (dart:async/zone.dart:1150:13)
#28     _CustomZone.runBinary (dart:async/zone.dart:1039:19)
#29     runZoned.<anonymous closure> (dart:async/zone.dart:1481:21)
#30     _CustomZone.handleUncaughtError (dart:async/zone.dart:1005:19)
#31     Future._propagateToListeners (dart:async/future_impl.dart:630:16)
#32     Future._completeError (dart:async/future_impl.dart:532:5)
#33     _AsyncAwaitCompleter.completeError (dart:async-patch/async_patch.dart:38:15)
#34     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart)
<asynchronous suspension>
#35     _rootRun (dart:async/zone.dart:1126:13)
#36     _CustomZone.run (dart:async/zone.dart:1023:19)
#37     _runZoned (dart:async/zone.dart:1518:10)
#38     runZoned (dart:async/zone.dart:1502:12)
#39     run.<anonymous closure> (package:flutter_tools/runner.dart:61:18)
<asynchronous suspension>
#40     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:146:29)
#41     _rootRun (dart:async/zone.dart:1126:13)
#42     _CustomZone.run (dart:async/zone.dart:1023:19)
#43     _runZoned (dart:async/zone.dart:1518:10)
#44     runZoned (dart:async/zone.dart:1465:12)
#45     AppContext.run (package:flutter_tools/src/base/context.dart:145:18)
#46     runInContext (package:flutter_tools/src/context_runner.dart:65:24)
#47     run (package:flutter_tools/runner.dart:50:10)
#48     main (package:flutter_tools/executable.dart:65:9)
#49     main (file:///Users/troysimon/development/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)
#50     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:305:32)
#51     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

If JSON is malformed don't crash tool

Fixes https://github.com/flutter/flutter/issues/45316

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/config.dart [hunk 1]
```diff
@@ -3,15 +3,23 @@
...
+  Config([File configFile, Logger localLogger]) {
+    final Logger loggerInstance = localLogger ?? logger;
     _configFile = configFile ?? fs.file(fs.path.join(userHomePath(), '.flutter_settings'));
     if (_configFile.existsSync()) {
-      _values = castStringKeyedMap(json.decode(_configFile.readAsStringSync()));
+      try {
+        _values = castStringKeyedMap(json.decode(_configFile.readAsStringSync()));
+      } on FormatException {
+        loggerInstance.printError('Failed to decode preferences in ${_configFile.path}.');
+        _configFile.deleteSync();
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/config.dart]: Should we delete this (I'm guessing probably, but wanted to check)? Are we assuming that the tool messed up, or are users manually editing these?

Assuming these are pretty small, should we at least log the contents of the corrupted config file?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/config.dart]: In the filled issue it looks like the file was corrupted - we could log the contents, but I'm worried about an edge case where it is accidentally a 2GB file.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/config.dart]: that's a good point about the edge case...

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/config.dart]: Do we persist analytics opt in data here?

**Review by darielopper — APPROVED** [review]

File: packages/flutter_tools/lib/src/base/config.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/config.dart]: The analytics opt in data is stored in $HOME/.flutter

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/config.dart]: I guess it's ok to blow this file away, then.

**Review by christopherfujino — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/base/config.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/config.dart]: I added more to the messaging regarding the deletion

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by altanseckin** [comment]
I have the same issue. Where is the .flutter_settings file. I am trying to delete it but i couldnt find it

**Comment by christopherfujino** [comment]
@altanseckin it should be in your user home path. On mac or linux, it should be at `$HOME/.flutter_settings`.

**Comment by altanseckin** [comment]
I work on win10 :D

**Comment by christopherfujino** [comment]
From the source code, the flutter tool uses this to determine `userHomePath`:

```dart
/// Reads the process environment to find the current user's home directory.
///
/// If the searched environment variables are not set, '.' is returned instead.
String userHomePath() {
  final String envKey = platform.operatingSystem == 'windows' ? 'APPDATA' : 'HOME';
  return platform.environment[envKey] ?? '.';
}
```

So if `APPDATA` is provided by your environment, the file will be created there. Else, it will be created wherever the tool was invoked from.

**Comment by altanseckin** [comment]
Thank you so much

**Pull request merged**